filegenerator/file_generator.go

```diff
@@ -0,0 +1,393 @@
+
+// BenchResults map dates to the benchmark test data for that date.
+type BenchResults map[string]BenchStats
+
+// BenchStats is a struct containing the various values obtained from running
+// a benchmark test with the -benchmem parameter. The variable names are kept
+// short to keep the JSON files small.
+type BenchStats struct {
+	N int     //nsop
+	A int     //alop
+	B int     //bop
+	M float64 //mbps
+}
+
+// GeometricMeanData is a struct that contains the means of the benchmark
+// stats for all the tests in a given package.
+type GeometricMeanData struct {
+	NMean float64
+	AMean float64
+	BMean float64
+	MMean float64
+	Date  string
+}
+
+func isValidTestLine(line string) bool {
+	matched, err := regexp.MatchString("FAIL", line)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return len(line) >= 9 && line[0:9] == "Benchmark" && !matched
+}
+
+func parseTestResultLine(line string) (string, BenchStats) {
+	testName := strings.Trim(strings.Split(line, "\t")[0], " ")
+	nsopr, _ := regexp.Compile("[0-9]+ ns/op")
+	allocpr, _ := regexp.Compile("[0-9]+ allocs/op")
+	bopr, _ := regexp.Compile("[0-9]+ B/op")
+	mbpsr, _ := regexp.Compile("[0-9]+\\.?[0-9]+ MB/s")
+	nsops := nsopr.FindString(line)
+	allocps := allocpr.FindString(line)
+	bops := bopr.FindString(line)
+	mbpss := mbpsr.FindString(line)
+	var nsopi, allocpi, bopi int
+	var mbpsf float64
+	if nsops != "" {
+		nsopi, _ = strconv.Atoi(strings.Split(nsops, " ")[0])
+	}
+	if allocps != "" {
+		allocpi, _ = strconv.Atoi(strings.Split(allocps, " ")[0])
+	}
+	if bops != "" {
+		bopi, _ = strconv.Atoi(strings.Split(bops, " ")[0])
+	}
+	if mbpss != "" {
+		mbpsf, _ = strconv.ParseFloat(strings.Split(mbpss, " ")[0], 64)
+	}
+	return testName, BenchStats{N: nsopi, A: allocpi, B: bopi, M: mbpsf}
+}
+
+func isValidDataDir(dataDirName string) bool {
+	matched, _ := regexp.MatchString("[0-9][0-9]-[0-9][0-9]-[0-9][0-9][0-9][0-9]", dataDirName)
+	return matched
+}
+
+// RenderHistoricalBenchmarkResults takes the folders in benchSamples, parses them, and returns a
+// datastructure called BenchPackages which represent the data from the folders.
+func RenderHistoricalBenchmarkResults(dirs []string) BenchPackages {
+	bPackages := make(map[string]BenchTestMap)
+	benchSamples := os.Getenv(benchSamplesEnv)
+	dataDirs, err := ioutil.ReadDir(benchSamples)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, dataDir := range dataDirs {
+		if !dataDir.IsDir() {
+			continue
+		}
+		dataDirName := dataDir.Name()
+		if !isValidDataDir(dataDirName) {
+			continue
+		}
+		for _, dirName := range dirs {
+			if _, ok := bPackages[dirName]; !ok {
+				bPackages[dirName] = make(BenchTestMap)
+			}
+			fullDirPath := benchSamples + "/" + dataDirName + "/cockroach/" + dirName
+			files, err := ioutil.ReadDir(fullDirPath)
+			if err != nil {
+				log.Fatal(err)
+			}
+			for _, file := range files {
+				fileName := file.Name()
+				matched, _ := regexp.MatchString(".*test\\.stdout", fileName)
```

Comment:
instead of using `"`, preferthe backtick quoting operator for regexes, to avoid having to excessively escape stuff like `\\`